internal/enum/enum.go

```diff
@@ -0,0 +1,353 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// enum is a helper for generating boilerplate related to Go enums.
+//
+// To generate boilerplate for a given file, use
+//
+//	//go:generate go run github.com/bufbuild/protocompile/internal/enum
+//
+// To generate code for an enum type, annotate it with the appropriate
+// directives.
+//
+//   - //enum:string Name will generate a stringifying methods named Name,
+//     which default to "String".
+//
+//   - //enum:gostring Name is similar, but the default name is "GoString"
+//     and each constant stringifies to "package.Constant".
+//
+//   - //enum:fromstring Name generates a function that takes a string and
+//     returns the constant that stringifies to that value, or 0 if there is
+//     no such value.
+//
+//   - //enum:doc directive "..." adds a line of doc comment to the method
+//     generated by directive with the given text.
+//
+//   - //enum:import "..." will include a package to import in the generated
+//     file.
+//
+//   - //enum:skip, attached to an enum constant, will cause that constant to
+//     be ignored by the generator.
+//
+//   - //enum:stringfunc Name provides a function to use for stringifying
+//     based on the name of a constant (see below).
```

Comment:
This functionality is used in only one place. I am so-so on whether it pulls its weight. Feel free to veto this.